### PR TITLE
Add option to deploy VM agents without a public IP [FIXED JENKINS-40620]

### PR DESCRIPTION
**Pull request by clguimanMSFT**

### CODE REVIEW

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java [hunk 1]
```diff
@@ -126,6 +125,8 @@
 
     private String subnetName;
 
+    private boolean usePrivateIp;
+
```
Comment (arroyc) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java]: nothing serious ... would be nice to see the name of the variable ends with "IP", but you can ignore it

Comment (clguimanMSFT) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java]: done

**Comment by mmitche** [comment]
Awesome.  I will get to looking at this this afternoon.

**Comment by clguimanMSFT** [comment]
@mmitche ping :)

File: src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java [hunk 2]
```diff
@@ -385,7 +458,16 @@ public static void setVirtualMachineDetails(
         VirtualMachine vm = azureClient.virtualMachines().getByGroup(template.getResourceGroupName(), azureAgent.getNodeName());
 
         // Getting the first virtual IP
-        azureAgent.setPublicDNSName(vm.getPrimaryPublicIpAddress().fqdn());
+        final PublicIpAddress publicIP = vm.getPrimaryPublicIpAddress();
+        String fqdn = "";
+        if (publicIP == null)
```
Comment (mmitche) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java]: Formatting, brace should be on same line as if

**Comment by clguimanMSFT** [comment]
@mmitche 
1. Yes, we try to connect to the public IP and if it's missing we're using the private one (https://github.com/jenkinsci/azure-vm-agents-plugin/pull/35/files#diff-b9ce3e407351f01915f57f0a146f7801R461). Of course that means that both master and agent have to be in the same vnet and subnet
2. The user can't easily ssh into the machine. He will either have to attach a public ip from the portal or have some network bridge from another machine he can reach.
I'm not really sure what we can write in the computer manangement page. I'll have a look

File: src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java (hunk 2 above)
Comment (clguimanMSFT) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java]: fixed

**Comment by mmitche** [comment]
@clguimanMSFT Yeah, I was questioning the private IP subnet thing.  Is there verification in here to ensure that the private IP and master are on the same vnet/snet?

We can write into the computer management page pretty easily.  It's one of the jelly files.  It would be interesting to have the master attach a PIP on demand.

**Comment by clguimanMSFT** [comment]
@mmitche The 'on-demand public ip attach' idea is very nice. I'll implement it
As for detecting if we can reach that private ip from master - we do that after the deployment was completed :) Before that we can't really know, because we don't know where the master is or if there are other machines in that subnet. 
In my TODO list there's a feature to show agent deployment issues in the Jenkins admin page (the red thing on top of the management page) so once we have that it will be straight forward for people to understand what's the problem with their template.
The private IP feature is opt-in so users will most likely be aware of what they have to do (I hope :) )

**Comment by mmitche** [comment]
@clguimanMSFT Like it, sounds good to me.

**Comment by clguimanMSFT** [comment]
@mmitche @arroyc 
I have added the feature suggested by Matt:

image

image

image

image

**Comment by scarvel8** [comment]
@clguimanMSFT:

Tested this myself, worked for me with a few modifications.

I am running this with a custom image template and a script:

1.) src/main/resources/customImageTemplateWithScript.json needs the same lines removed as the other files:
                            "publicIPAddress": {
                                 "id": "[resourceId('Microsoft.Network/publicIPAddresses', concat(variables('vmName'), copyIndex(), 'IPName'))]"
                              },

2.) There is no need to restrict vnetID in customImageTemplateWithScript to the resource group, I am using a network interface that is outside of the resource group in the same subscription, just need to specify "/subscriptions/(subscriptionid)/resourceGroups/(resourcegroupofnic)/providers/Microsoft.Network/virtualNetworks/(virtualnetwork)". I took out the whole if block `if (StringUtils.isNotBlank(virtualNetworkName))` in AzureVMManagementServiceDelegate.java for validation to pass and the deployment to trigger.

3.) The password I wanted to use didn't pass validation, make isValidPassword always return true (not sure why it's there in the first place).

Would be great if these can be ironed out, I quickly hacked it personally just to POC this, but would like to see #1 & #2 integrated into the plugin.

**Comment by clguimanMSFT** [comment]
@scarvel8 
#1 - Nice catch. I updated the commit
#2 - That sounds like a separate feature (requires some UI work as well). Please create a feature request here: https://issues.jenkins-ci.org/browse/JENKINS-43243?jql=project%20%3D%20JENKINS%20AND%20component%20%3D%20azure-vm-agents-plugin
#3 It was already fixed by https://github.com/jenkinsci/azure-vm-agents-plugin/commit/957d2b265b1e89d35b1d5d6564367e78d3dfdf86

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java [hunk 3]
```diff
@@ -435,6 +457,26 @@ public boolean isVMAliveOrHealthy() throws Exception {
...
+    private final Object publicIPAttachLock = new Object();
+
+    public String attachPublicIP() {
+        if (!publicIP.isEmpty()) {
+            return publicIP;
+        }
+        //attachPublicIP will try and provision a public IP or will return if one already exists
+        // because of that we can just wait here until we have a public IP
+        synchronized (publicIPAttachLock) {
+            AzureVMCloud azureVMCloud = (AzureVMCloud) Jenkins.getInstance().getCloud(cloudName);
```
Comment (mmitche) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java]: You might want to move this inside the try.  It's probably not critical, but deletion of the parent cloud (or change of name, etc.) will leave existing nodes around but getCloud will fail.

File: src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java [hunk 4]
```diff
@@ -385,14 +458,53 @@ public static void setVirtualMachineDetails(
...
+            if (vm.getPrimaryPublicIpAddress() == null) {
+                vm.getPrimaryNetworkInterface()
+                        .update()
+                        .withNewPrimaryPublicIpAddress(
+                                azureClient.publicIpAddresses()
+                                        .define(azureAgent.getNodeName() + "IPName")
+                                        .withRegion(template.getLocation())
+                                        .withExistingResourceGroup(template.getResourceGroupName())
+                                        .withLeafDomainLabel(azureAgent.getNodeName())
+                        ).apply();
```
Comment (mmitche) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java]: I assume there isn't a race here?  When this returns the VM details sent back from Azure will show the PIP?

Comment (clguimanMSFT) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java]: apply() is synchronous, it doesn't return until the PIP is actually created. 
setVirtualMachineDetails asks Azure again so it should always get the PIP (I've tested empirically and it does - if it wont' then it's an SDK bug)

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java (hunk 3 above)
Comment (clguimanMSFT) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java]: good point. will do

**Pull request merged**